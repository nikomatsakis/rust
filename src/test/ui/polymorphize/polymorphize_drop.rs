// Simple test for the polymorphize analysis code.
//
// compile-flags: -Zpolymorphize -Zpolymorphize-dump
// compile-pass

#![allow(warnings)]

struct OffsetDependent<T> {
    t: T,
    count: u32,
}

fn dependency_because_drop_T<T>(data: T) {
    //~^ ERROR some polymorphic dependencies found
}

fn dependency_because_drop_box_T<T>(data: Box<T>) {
    //~^ ERROR some polymorphic dependencies found
}

fn no_dependency_because_drop_ref_T<T>(data: &T) {
    //~^ ERROR no polymorphic dependencies found
}

fn no_dependency_because_indirect_drop_ref_T<T>(data: &T) {
    //~^ ERROR no polymorphic dependencies found

    dependency_because_drop_T::<&T>(data);
}

fn no_dependency_because_indirect_drop_ref_u32() {
    //~^ ERROR no polymorphic dependencies found

    dependency_because_drop_T::<&u32>(&22);
}

fn main() {
    //~^ ERROR no polymorphic dependencies found
}
