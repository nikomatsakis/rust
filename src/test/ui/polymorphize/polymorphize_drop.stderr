warning: some polymorphic dependencies found
  --> $DIR/polymorphize_drop.rs:13:1
   |
LL | / fn dependency_because_drop_T<T>(data: T) {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | | }
   | | ^
   | | |
   | |_depends on `Drop(T)`
   | 

warning: some polymorphic dependencies found
  --> $DIR/polymorphize_drop.rs:17:1
   |
LL | / fn dependency_because_drop_box_T<T>(data: Box<T>) {
LL | |     //~^ ERROR some polymorphic dependencies found
LL | | }
   | | ^
   | | |
   | |_depends on `Drop(std::boxed::Box<T>)`
   | 

warning: no polymorphic dependencies found
  --> $DIR/polymorphize_drop.rs:21:1
   |
LL | / fn no_dependency_because_drop_ref_T<T>(data: &T) {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | | }
   | |_^

warning: no polymorphic dependencies found
  --> $DIR/polymorphize_drop.rs:25:1
   |
LL | / fn no_dependency_because_indirect_drop_ref_T<T>(data: &T) {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | |     dependency_because_drop_T::<&T>(data);
   | |     ------------------------------------- invokes `DefId(0/0:4 ~ polymorphize_drop[317d]::dependency_because_drop_T[0])` with substitutions `[&T]`
LL | | }
   | |_^

warning: no polymorphic dependencies found
  --> $DIR/polymorphize_drop.rs:30:1
   |
LL | / fn no_dependency_because_indirect_drop_ref_u32() {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | |     dependency_because_drop_T::<&u32>(&22);
   | |     -------------------------------------- invokes `DefId(0/0:4 ~ polymorphize_drop[317d]::dependency_because_drop_T[0])` with substitutions `[&u32]`
LL | | }
   | |_^

warning: no polymorphic dependencies found
  --> $DIR/polymorphize_drop.rs:35:1
   |
LL | / fn main() {
LL | |     //~^ ERROR no polymorphic dependencies found
LL | | }
   | |_^

note: Monomorphized items: 9

note: Monomorphized size : 32

note: New total items    : 9 (100%)

note: New estimated size : 32 (100%)

